src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c

```diff
@@ -1658,21 +1677,20 @@ static void lb_on_response_received_locked(grpc_exec_ctx *exec_ctx, void *arg,
             }
             grpc_grpclb_destroy_serverlist(serverlist);
           } else { /* new serverlist */
-            /* record the old status */
-            /* TODO(juanlishen): this assumes that num_servers never decreases
-             * to 0 once it's positive */
-            bool first_serverlist = glb_policy->serverlist == NULL;
             if (glb_policy->serverlist != NULL) {
               /* dispose of the old serverlist */
               grpc_grpclb_destroy_serverlist(glb_policy->serverlist);
-            } 
+            }
             /* and update the copy in the glb_lb_policy instance. This
              * serverlist instance will be destroyed either upon the next
              * update or in glb_destroy() */
             glb_policy->serverlist = serverlist;
             glb_policy->serverlist_index = 0;
-            if (first_serverlist) {
-              /* Don't use fallback any more. */
+            /* dispose the fallback (if any). It should only be done once when the 
+             * first non-empty \a serverlist is received */
+            if (glb_policy->fallback_backend_addresses != NULL) {
```

Comment:
This can be moved up to an `else` clause on line 1683 above.

Reply:
Yes, these two conditions are mutual exclusive. But I separated them because
- the processing of the `serverlist` and the processing of the fallback can be clearly separated;
- the fallback addresses can be destroyed just after the `glb_policy` has a non-empty `serverlist` (this order doesn't improve any correctness, but it's more consistent with (my) intuition);
- this is consistent with the resolution update code on line 1896 (although it can also be changed to use `serverlist` to check).

Reply:
> - the processing of the `serverlist` and the processing of the fallback can be clearly separated;

As per my comment elsewhere, I think the code currently guarantees that exactly one of `serverlist` and `fallback_backend_addresses` will be non-NULL, and I think we should be consistent about using a single criteria to determine which one should be set.

> - the fallback addresses can be destroyed just after the `glb_policy` has a non-empty `serverlist` (this order doesn't improve any correctness, but it's more consistent with (my) intuition);

As you say, this doesn't improve correctness, so I don't think this is an issue.  It's important to understand the locking properties of this code, and I think your intuition needs to take that into account. :)

> - this is consistent with the resolution update code on line 1896 (although it can also be changed to use `serverlist` to check).

As I mentioned there, I think we should be consistent about this.  I think both places should look at `serverlist`.